crates/astria-sequencer/src/accounts/transaction.rs

```diff
@@ -0,0 +1,72 @@
+use anyhow::{
+    ensure,
+    Result,
+};
+use async_trait::async_trait;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+
+use crate::{
+    accounts::{
+        state_ext::{
+            StateReadExt,
+            StateWriteExt,
+        },
+        types::{
+            Address,
+            Balance,
+            Nonce,
+        },
+    },
+    transaction::ActionHandler,
+};
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+pub struct Transaction {
+    to: Address,
+    from: Address,
+    amount: Balance,
+    nonce: Nonce,
+}
+
+impl Transaction {
+    pub fn new(to: Address, from: Address, amount: Balance, nonce: Nonce) -> Self {
+        Self {
+            to,
+            from,
+            amount,
+            nonce,
+        }
+    }
+}
+
+#[async_trait]
+impl ActionHandler for Transaction {
+    fn check_stateless(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn check_stateful<S: StateReadExt + 'static>(&self, state: &S) -> Result<()> {
+        let curr_nonce = state.get_account_nonce(&self.from).await?;
+
+        // TODO: do nonces start at 0 or 1? this assumes an account's first tx has nonce 1.
+        ensure!(curr_nonce < self.nonce, "invalid nonce",);
+
+        let curr_balance = state.get_account_balance(&self.from).await?;
+        ensure!(curr_balance >= self.amount, "insufficient funds",);
+
+        Ok(())
+    }
+
+    async fn execute<S: StateWriteExt>(&self, state: &mut S) -> Result<()> {
+        let from_balance = state.get_account_balance(&self.from).await?;
+        let from_nonce = state.get_account_nonce(&self.from).await?;
+        let to_balance = state.get_account_balance(&self.to).await?;
+        state.put_account_balance(&self.from, from_balance - self.amount)?;
```

Comment:
Maybe this is just that lower level, but it feels like maybe the logic for state transitions which has race possibilities should be handled by the state itself? If putting the balance succeeds but nonce fails, we have issues.

Reply:
right, so these writes are all happening on a `StateDelta`, which is basically a list of pending writes that happen during a block's execution. the writes are all batched and only actually written to state once `Commit` is called. so yeah this effectively is atomic afaik

Reply:
Can we create an issue to add tests for error cases here?